### ISSUE TITLE
kvs: implement atomic append

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#1193]
The ability to append to a KVS value without needing to perform a read-modify-write is useful for the following use cases:
* stdio - the "libkz" mechanism is an elaborate workaround for the lack of an atomic append
* state logs - a "log" stored in a KVS value has been proposed for tracking job state changes; atomicity is needed if there are to be multiple writers.

Implementation notes:  this can look exactly like a "put" from the point of view of a client, e.g. there could be a `flux_kvs_txn_append()` function which adds a new append operation to a transaction.  From the point of view of the KVS module, commit processing would need a new operation which would add another content reference to the array in a _valref_ object (possibly converting a _val_ to a _valref_ on first append).

If libkz is to be replaced, we need a way to efficiently "tail" a value that's being appended to.  One idea is to implement a "get" mechanism to retrieve the nth blob of a _valref_, that optionally defers the response until that blob exists.  Another is to "watch" the _valref_, keeping track of current position in its blobref array.  Retaining the integrity of each append as a distinct blob, rather than trying to pack them for space efficiency, would be required for either method.  This property does open some interesting possibilities for a sort of "object stream" model where the blobs each contain a valid JSON object or or log line or whatever.

**Comment by chu11** [comment]
Assuming the ```lookup``` in #1202 is completed first, I see this issue having three parts.

1) Support the actual append operation.  Will include function such as ```flux_kvs_txn_append()``` and the KVS server side needs listed above.  A ```flux-kvs``` equivalent will be needed as well, which will not look that much different than ```flux kvs put```.  Lookups on such values will return the aggregate value as described in #1202.

2) Support new ```lookup``` operations for blobrefs.  Exact needs TBD, dependent on API needs from libkz or other libraries.  This can be designed after #1 above is complete.  Some ideas (including some listed above), each of which could be separate PRs.
- "get" mechanism to retrieve nth blob
- "get" mechanism to retrieve ith-jth blob (i.e. ```tail -n 5``` like support)
- "watch" with tracking position in blobref array
- "watch" returning only most recently appended
- unclear which of the above will need ```flux-kvs``` support (new command or options to ```get```/```watch```), but support somewhere (e.g. ```t/kvs/basic```) for testing will be needed.

3) updates to libkz and other libraries.  This is really for a separate issue.  But thought it should be made clear this is to be spliced out from the above.

**Comment by garlick** [comment]
On implementing the append API and wire protocol:

The wire protocol for commit/fence will need to change to support append.  The request currently includes an array of operations, each of which consists of 
* key (fully qualified key name)
* treeobj (any RFC 11 tree object, e.g.  a _val_ object)

I propose we add to each operation
* flags (integer)

The `flux_kvs_txn_` functions already support a _flags_ parameter that is unused.  We could add a flag FLUX_KVS_APPEND, and then one would append by calling
```c
flux_kvs_txn_put_raw (txn, FLUX_KVS_APPEND, key, data, len);
```

If supplied to the other functions I guess it would be an EINVAL error, at least initially.

**Comment by chu11** [comment]
Its worth mentioning that when implementing append, there are really two types of append

1) the server receives data directly (internally via a val tree object)
2) the server receives blobref(s) (internally via a valref tree object)

The first is the more common append one would think about.  The latter is a performance optimization for large data.  The data is stored in the content store in a separate operation (via direct writing to the content store or prior KVS operations) and the user is aware of the blobref value(s).  Instead of passing the entire value to the KVS server for append, the user can pass just the blobref(s) to the KVS server.

The latter is a performance optimization that may not be implemented initially.

**Comment by garlick** [comment]
Agreed.

I kind of lump case 2 in with the "committing large values" section of #1202 which we decided to defer.

If you decide to support it on the server (since seems to be a subset of case 1 on the server end), we could make up test cases involving
```c
flux_kvs_txn_put_treeobj (FLUX_KVS_APPEND, valref)
```
or
```
flux kvs put --append --treeobj ...
```

**Comment by garlick** [comment]
With #1265 merged, we have support for an atomic append operation.  I'll close this and we can add new issues for other things we want when we decide for sure we want them.
